pkg/sinks/prometheus.go

```diff
@@ -0,0 +1,104 @@
+package sinks
+
+import (
+	"context"
+	"strings"
+
+	"k8s.io/utils/strings/slices"
+
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/resmoio/kubernetes-event-exporter/pkg/kube"
+	"github.com/rs/zerolog/log"
+)
+
+func newGaugeVec(opts prometheus.GaugeOpts, labelNames []string) *prometheus.GaugeVec {
+	v := prometheus.NewGaugeVec(opts, labelNames)
+	prometheus.MustRegister(v)
+	return v
+}
+
+type PrometheusConfig struct {
+	EventsMetricsNamePrefix string              `yaml:"eventsMetricsNamePrefix"`
+	ReasonFilter            map[string][]string `yaml:"reasonFilter"`
+}
+
+type PrometheusGaugeVec interface {
+	With(labels prometheus.Labels) prometheus.Gauge
+	Delete(labels prometheus.Labels) bool
+}
+
+type PrometheusSink struct {
+	cfg           *PrometheusConfig
+	kinds         []string
+	metricsByKind map[string]PrometheusGaugeVec
+}
+
+func NewPrometheusSink(config *PrometheusConfig) (Sink, error) {
+	if config.EventsMetricsNamePrefix == "" {
+		config.EventsMetricsNamePrefix = "event_exporter_"
+	}
+
+	metricsByKind := map[string]PrometheusGaugeVec{}
+
+	log.Info().Msgf("Initializing new Prometheus sink...")
+	kinds := []string{}
+	for kind := range config.ReasonFilter {
+		kinds = append(kinds, kind)
+		metricName := config.EventsMetricsNamePrefix + strings.ToLower(kind) + "_event_count"
+		metricLabels := []string{strings.ToLower(kind), "namespace", "reason"}
+		metricsByKind[kind] = newGaugeVec(
+			prometheus.GaugeOpts{
+				Name: metricName,
+				Help: "Event counts for " + kind + " resources.",
+			}, metricLabels)
+
+		log.Info().Msgf("Created metric: %s, will emit events: %v with additional labels: %v", kind, config.ReasonFilter[kind], metricLabels)
+	}
+
+	return &PrometheusSink{
+		cfg:           config,
+		kinds:         kinds,
+		metricsByKind: metricsByKind,
+	}, nil
+}
+
+func (o *PrometheusSink) Send(ctx context.Context, ev *kube.EnhancedEvent) error {
+	kind := ev.InvolvedObject.Kind
+	if slices.Contains(o.kinds, kind) {
+		for _, reason := range o.cfg.ReasonFilter[kind] {
```

Comment:
I've tested this locally and saw that if you have a config like this

```yaml
logLevel: info
logFormat: json
route:
  routes:
    - match:
        - receiver: "prometheus"
receivers:
  - name: "prometheus"
    prometheus:
      eventsMetricsNamePrefix: "my_test_prefix_"
      reasonFilter:
        Pod:
          - "Scheduled"
          - "Created"
```

and i have the following events:

```sh
12h         Normal    Created             pod/coredns-7889b4b8fd-phgf6                     Created container coredns
12h         Normal    Scheduled           pod/coredns-7889b4b8fd-phgf6                     Successfully assigned kube-system/coredns-7889b4b8fd-phgf6 to psql-control-plane
12h         Normal    Scheduled           pod/coredns-7889b4b8fd-qdc72                     Successfully assigned kube-system/coredns-7889b4b8fd-qdc72 to psql-control-plane
12h         Normal    Created             pod/coredns-7889b4b8fd-qdc72                     Created container coredns
12h         Normal    Scheduled           pod/kindnet-t6j2f                                Successfully assigned kube-system/kindnet-t6j2f to psql-control-plane
12h         Normal    Scheduled           pod/kube-proxy-xw9bc                             Successfully assigned kube-system/kube-proxy-xw9bc to psql-control-plane
12h         Normal    Created             pod/kindnet-t6j2f                                Created container kindnet-cni
12h         Normal    Created             pod/kube-proxy-xw9bc                             Created container kube-proxy
12h         Normal    Scheduled           pod/coredns-655f47cb8b-ppd6f                     Successfully assigned kube-system/coredns-655f47cb8b-ppd6f to psql-control-plane
12h         Normal    Scheduled           pod/coredns-655f47cb8b-vv2ng                     Successfully assigned kube-system/coredns-655f47cb8b-vv2ng to psql-control-plane
12h         Normal    Created             pod/coredns-655f47cb8b-ppd6f                     Created container coredns
12h         Normal    Created             pod/coredns-655f47cb8b-vv2ng                     Created container coredns
12h         Normal    Scheduled           pod/kube-proxy-84w5m                             Successfully assigned kube-system/kube-proxy-84w5m to psql-control-plane
12h         Normal    Scheduled           pod/kindnet-hmdh4                                Successfully assigned kube-system/kindnet-hmdh4 to psql-control-plane
12h         Normal    Created             pod/kube-proxy-84w5m                             Created container kube-proxy
12h         Normal    Created             pod/kindnet-hmdh4                                Created container kindnet-cni
12h         Normal    Scheduled           pod/kindnet-q79dm                                Successfully assigned kube-system/kindnet-q79dm to psql-control-plane
12h         Normal    Scheduled           pod/kube-proxy-2tcjr                             Successfully assigned kube-system/kube-proxy-2tcjr to psql-control-plane
12h         Normal    Created             pod/kindnet-q79dm                                Created container kindnet-cni
12h         Normal    Created             pod/kube-proxy-2tcjr                             Created container kube-proxy
12h         Normal    Scheduled           pod/kindnet-ljxln                                Successfully assigned kube-system/kindnet-ljxln to psql-control-plane
12h         Normal    Scheduled           pod/kube-proxy-hrrv7                             Successfully assigned kube-system/kube-proxy-hrrv7 to psql-control-plane
12h         Normal    Created             pod/kindnet-ljxln                                Created container kindnet-cni
12h         Normal    Created             pod/kube-proxy-hrrv7                             Created container kube-proxy
12h         Normal    Scheduled           pod/kindnet-zt95n                                Successfully assigned kube-system/kindnet-zt95n to psql-control-plane
12h         Normal    Created             pod/kindnet-zt95n                                Created container kindnet-cni
12h         Normal    Scheduled           pod/kube-proxy-qkghc                             Successfully assigned kube-system/kube-proxy-qkghc to psql-control-plane
12h         Normal    Created             pod/kube-proxy-qkghc                             Created container kube-proxy
15m         Normal    Created             pod/etcd-psql-control-plane                      Created container etcd
15m         Normal    Created             pod/kube-controller-manager-psql-control-plane   Created container kube-controller-manager
15m         Normal    Created             pod/kube-apiserver-psql-control-plane            Created container kube-apiserver
15m         Normal    Created             pod/kube-scheduler-psql-control-plane            Created container kube-scheduler
15m         Normal    Created             pod/coredns-655f47cb8b-ppd6f                     Created container coredns
15m         Normal    Created             pod/kube-proxy-qkghc                             Created container kube-proxy
15m         Normal    Created             pod/coredns-655f47cb8b-vv2ng                     Created container coredns
15m         Normal    Created             pod/kindnet-zt95n                                Created container kindnet-cni
12m         Normal    Scheduled           pod/kindnet-w9d5z                                Successfully assigned kube-system/kindnet-w9d5z to psql-control-plane
12m         Normal    Scheduled           pod/kube-proxy-w2lmw                             Successfully assigned kube-system/kube-proxy-w2lmw to psql-control-plane
12m         Normal    Created             pod/kindnet-w9d5z                                Created container kindnet-cni
12m         Normal    Created             pod/kube-proxy-w2lmw                             Created container kube-proxy
12m         Normal    Scheduled           pod/coredns-7fb4cf5c7f-6q4tk                     Successfully assigned kube-system/coredns-7fb4cf5c7f-6q4tk to psql-control-plane
12m         Normal    Scheduled           pod/coredns-7fb4cf5c7f-l97hc                     Successfully assigned kube-system/coredns-7fb4cf5c7f-l97hc to psql-control-plane
12m         Normal    Created             pod/coredns-7fb4cf5c7f-6q4tk                     Created container coredns
12m         Normal    Created             pod/coredns-7fb4cf5c7f-l97hc                     Created container coredns
7m29s       Normal    Scheduled           pod/coredns-5fc45464b4-x2rg5                     Successfully assigned kube-system/coredns-5fc45464b4-x2rg5 to psql-control-plane
7m29s       Normal    Scheduled           pod/coredns-5fc45464b4-gz5dh                     Successfully assigned kube-system/coredns-5fc45464b4-gz5dh to psql-control-plane
7m29s       Normal    Created             pod/coredns-5fc45464b4-gz5dh                     Created container coredns
7m29s       Normal    Created             pod/coredns-5fc45464b4-x2rg5                     Created container coredns
2m39s       Normal    Scheduled           pod/coredns-cb668d945-4rh2m                      Successfully assigned kube-system/coredns-cb668d945-4rh2m to psql-control-plane
2m39s       Normal    Created             pod/coredns-cb668d945-jgglt                      Created container coredns
2m39s       Normal    Scheduled           pod/coredns-cb668d945-jgglt                      Successfully assigned kube-system/coredns-cb668d945-jgglt to psql-control-plane
2m39s       Normal    Created             pod/coredns-cb668d945-4rh2m                      Created container coredns
```

and the logs give me something like this:

```sh
{"level":"info","time":"2024-06-18T09:23:03+02:00","message":"evReason Scheduled and reason Scheduled"}
{"level":"info","time":"2024-06-18T09:23:03+02:00","message":"Setting event count metric with labels: map[namespace:kube-system pod:coredns-cb668d945-jgglt reason:Scheduled]"}
{"level":"info","time":"2024-06-18T09:23:03+02:00","message":"evReason Scheduled and reason Created"}
{"level":"info","time":"2024-06-18T09:23:03+02:00","message":"Deleting event count metric with labels: map[namespace:kube-system pod:coredns-cb668d945-jgglt reason:Created], and reason: \"Created\""}
{"level":"error","error":"pods \"coredns-5fc45464b4-gz5dh\" not found","time":"2024-06-18T09:23:08+02:00","message":"Object not found, likely deleted"}
{"level":"info","time":"2024-06-18T09:23:08+02:00","message":"evReason Unhealthy and reason Scheduled"}
{"level":"info","time":"2024-06-18T09:23:08+02:00","message":"Deleting event count metric with labels: map[namespace:kube-system pod:coredns-5fc45464b4-gz5dh reason:Scheduled], and reason: \"Scheduled\""}
{"level":"info","time":"2024-06-18T09:23:08+02:00","message":"evReason Unhealthy and reason Created"}
{"level":"info","time":"2024-06-18T09:23:08+02:00","message":"Deleting event count metric with labels: map[namespace:kube-system pod:coredns-5fc45464b4-gz5dh reason:Created], and reason: \"Created\""}
I0618 09:23:11.188323    4593 request.go:690] Waited for 1.797702008s due to client-side throttling, not priority and fairness, request: GET:https://127.0.0.1:50947/apis/networking.k8s.io/v1
{"level":"info","time":"2024-06-18T09:23:14+02:00","message":"evReason Created and reason Scheduled"}
{"level":"info","time":"2024-06-18T09:23:14+02:00","message":"Deleting event count metric with labels: map[namespace:kube-system pod:coredns-cb668d945-4rh2m reason:Scheduled], and reason: \"Scheduled\""}
{"level":"info","time":"2024-06-18T09:23:14+02:00","message":"evReason Created and reason Created"}
{"level":"info","time":"2024-06-18T09:23:14+02:00","message":"Setting event count metric with labels: map[namespace:kube-system pod:coredns-cb668d945-4rh2m reason:Created]"}
{"level":"info","time":"2024-06-18T09:23:14+02:00","message":"evReason Pulled and reason Scheduled"}
{"level":"info","time":"2024-06-18T09:23:14+02:00","message":"Deleting event count metric with labels: map[namespace:kube-system pod:coredns-cb668d945-4rh2m reason:Scheduled], and reason: \"Scheduled\""}
{"level":"info","time":"2024-06-18T09:23:14+02:00","message":"evReason Pulled and reason Created"}
{"level":"info","time":"2024-06-18T09:23:14+02:00","message":"Deleting event count metric with labels: map[namespace:kube-system pod:coredns-cb668d945-4rh2m reason:Created], and reason: \"Created\""}
{"level":"info","time":"2024-06-18T09:23:19+02:00","message":"evReason Pulled and reason Scheduled"}
{"level":"info","time":"2024-06-18T09:23:19+02:00","message":"Deleting event count metric with labels: map[namespace:kube-system pod:coredns-cb668d945-jgglt reason:Scheduled], and reason: \"Scheduled\""}
{"level":"info","time":"2024-06-18T09:23:19+02:00","message":"evReason Pulled and reason Created"}
{"level":"info","time":"2024-06-18T09:23:19+02:00","message":"Deleting event count metric with labels: map[namespace:kube-system pod:coredns-cb668d945-jgglt reason:Created], and reason: \"Created\""}
{"level":"info","time":"2024-06-18T09:23:19+02:00","message":"evReason Created and reason Scheduled"}
{"level":"info","time":"2024-06-18T09:23:19+02:00","message":"Deleting event count metric with labels: map[namespace:kube-system pod:coredns-cb668d945-jgglt reason:Scheduled], and reason: \"Scheduled\""}
{"level":"info","time":"2024-06-18T09:23:19+02:00","message":"evReason Created and reason Created"}
{"level":"info","time":"2024-06-18T09:23:19+02:00","message":"Setting event count metric with labels: map[namespace:kube-system pod:coredns-cb668d945-jgglt reason:Created]"}
{"level":"info","time":"2024-06-18T09:23:19+02:00","message":"evReason Started and reason Scheduled"}
{"level":"info","time":"2024-06-18T09:23:19+02:00","message":"Deleting event count metric with labels: map[namespace:kube-system pod:coredns-cb668d945-jgglt reason:Scheduled], and reason: \"Scheduled\""}
{"level":"info","time":"2024-06-18T09:23:19+02:00","message":"evReason Started and reason Created"}
{"level":"info","time":"2024-06-18T09:23:19+02:00","message":"Deleting event count metric with labels: map[namespace:kube-system pod:coredns-cb668d945-jgglt reason:Created], and reason: \"Created\""}
{"level":"info","time":"2024-06-18T09:23:19+02:00","message":"evReason Started and reason Scheduled"}
{"level":"info","time":"2024-06-18T09:23:19+02:00","message":"Deleting event count metric with labels: map[namespace:kube-system pod:coredns-cb668d945-4rh2m reason:Scheduled], and reason: \"Scheduled\""}
{"level":"info","time":"2024-06-18T09:23:19+02:00","message":"evReason Started and reason Created"}
{"level":"info","time":"2024-06-18T09:23:19+02:00","message":"Deleting event count metric with labels: map[namespace:kube-system pod:coredns-cb668d945-4rh2m reason:Created], and reason: \"Created\""}
I0618 09:23:21.387313    4593 request.go:690] Waited for 1.596645393s due to client-side throttling, not priority and fairness, request: GET:https://127.0.0.1:50947/apis/batch/v1
{"level":"error","error":"pods \"coredns-5fc45464b4-x2rg5\" not found","time":"2024-06-18T09:23:29+02:00","message":"Object not found, likely deleted"}
{"level":"info","time":"2024-06-18T09:23:29+02:00","message":"evReason Killing and reason Scheduled"}
{"level":"info","time":"2024-06-18T09:23:29+02:00","message":"Deleting event count metric with labels: map[namespace:kube-system pod:coredns-5fc45464b4-x2rg5 reason:Scheduled], and reason: \"Scheduled\""}
{"level":"info","time":"2024-06-18T09:23:29+02:00","message":"evReason Killing and reason Created"}
{"level":"info","time":"2024-06-18T09:23:29+02:00","message":"Deleting event count metric with labels: map[namespace:kube-system pod:coredns-5fc45464b4-x2rg5 reason:Created], and reason: \"Created\""}
```

I'm wondering what the expected metrics would look like.
If i open the metrics page then i see only this:

```sh
# HELP build_info A metric with a constant '1' value labeled by version, revision, branch, and goversion from which Kubernetes Event Exporter was built.
# TYPE build_info gauge
build_info{goarch="amd64",goos="darwin",goversion="go1.22.0",revision="f9cd7cd472cec6311438f68ef4fa817fa6f28238",version="unknown"} 1
# HELP events_discarded The total number of events discarded because of being older than the maxEventAgeSeconds specified
# TYPE events_discarded counter
events_discarded 0
# HELP events_sent The total number of events processed
# TYPE events_sent counter
events_sent 17
# HELP go_build_info Build information about the main Go module.
# TYPE go_build_info gauge
go_build_info{checksum="",path="github.com/resmoio/kubernetes-event-exporter",version="(devel)"} 1
# HELP go_gc_duration_seconds A summary of the pause duration of garbage collection cycles.
# TYPE go_gc_duration_seconds summary
go_gc_duration_seconds{quantile="0"} 4.8081e-05
go_gc_duration_seconds{quantile="0.25"} 5.2652e-05
go_gc_duration_seconds{quantile="0.5"} 6.1909e-05
go_gc_duration_seconds{quantile="0.75"} 0.000107455
go_gc_duration_seconds{quantile="1"} 0.000184886
go_gc_duration_seconds_sum 0.000761977
go_gc_duration_seconds_count 9
# HELP go_goroutines Number of goroutines that currently exist.
# TYPE go_goroutines gauge
go_goroutines 23
# HELP go_info Information about the Go environment.
# TYPE go_info gauge
go_info{version="go1.22.0"} 1
# HELP go_memstats_alloc_bytes Number of bytes allocated and still in use.
# TYPE go_memstats_alloc_bytes gauge
go_memstats_alloc_bytes 7.625904e+06
# HELP go_memstats_alloc_bytes_total Total number of bytes allocated, even if freed.
# TYPE go_memstats_alloc_bytes_total counter
go_memstats_alloc_bytes_total 4.186132e+07
# HELP go_memstats_buck_hash_sys_bytes Number of bytes used by the profiling bucket hash table.
# TYPE go_memstats_buck_hash_sys_bytes gauge
go_memstats_buck_hash_sys_bytes 1.465109e+06
# HELP go_memstats_frees_total Total number of frees.
# TYPE go_memstats_frees_total counter
go_memstats_frees_total 162961
# HELP go_memstats_gc_sys_bytes Number of bytes used for garbage collection system metadata.
# TYPE go_memstats_gc_sys_bytes gauge
go_memstats_gc_sys_bytes 3.488312e+06
# HELP go_memstats_heap_alloc_bytes Number of heap bytes allocated and still in use.
# TYPE go_memstats_heap_alloc_bytes gauge
go_memstats_heap_alloc_bytes 7.625904e+06
# HELP go_memstats_heap_idle_bytes Number of heap bytes waiting to be used.
# TYPE go_memstats_heap_idle_bytes gauge
go_memstats_heap_idle_bytes 1.2689408e+07
# HELP go_memstats_heap_inuse_bytes Number of heap bytes that are in use.
# TYPE go_memstats_heap_inuse_bytes gauge
go_memstats_heap_inuse_bytes 1.1051008e+07
# HELP go_memstats_heap_objects Number of allocated objects.
# TYPE go_memstats_heap_objects gauge
go_memstats_heap_objects 27092
# HELP go_memstats_heap_released_bytes Number of heap bytes released to OS.
# TYPE go_memstats_heap_released_bytes gauge
go_memstats_heap_released_bytes 5.455872e+06
# HELP go_memstats_heap_sys_bytes Number of heap bytes obtained from system.
# TYPE go_memstats_heap_sys_bytes gauge
go_memstats_heap_sys_bytes 2.3740416e+07
# HELP go_memstats_last_gc_time_seconds Number of seconds since 1970 of last garbage collection.
# TYPE go_memstats_last_gc_time_seconds gauge
go_memstats_last_gc_time_seconds 1.718695417245406e+09
# HELP go_memstats_lookups_total Total number of pointer lookups.
# TYPE go_memstats_lookups_total counter
go_memstats_lookups_total 0
# HELP go_memstats_mallocs_total Total number of mallocs.
# TYPE go_memstats_mallocs_total counter
go_memstats_mallocs_total 190053
# HELP go_memstats_mcache_inuse_bytes Number of bytes in use by mcache structures.
# TYPE go_memstats_mcache_inuse_bytes gauge
go_memstats_mcache_inuse_bytes 7200
# HELP go_memstats_mcache_sys_bytes Number of bytes used for mcache structures obtained from system.
# TYPE go_memstats_mcache_sys_bytes gauge
go_memstats_mcache_sys_bytes 15600
# HELP go_memstats_mspan_inuse_bytes Number of bytes in use by mspan structures.
# TYPE go_memstats_mspan_inuse_bytes gauge
go_memstats_mspan_inuse_bytes 194720
# HELP go_memstats_mspan_sys_bytes Number of bytes used for mspan structures obtained from system.
# TYPE go_memstats_mspan_sys_bytes gauge
go_memstats_mspan_sys_bytes 244800
# HELP go_memstats_next_gc_bytes Number of heap bytes when next garbage collection will take place.
# TYPE go_memstats_next_gc_bytes gauge
go_memstats_next_gc_bytes 1.5889144e+07
# HELP go_memstats_other_sys_bytes Number of bytes used for other system allocations.
# TYPE go_memstats_other_sys_bytes gauge
go_memstats_other_sys_bytes 1.214091e+06
# HELP go_memstats_stack_inuse_bytes Number of bytes in use by the stack allocator.
# TYPE go_memstats_stack_inuse_bytes gauge
go_memstats_stack_inuse_bytes 1.376256e+06
# HELP go_memstats_stack_sys_bytes Number of bytes obtained from system for stack allocator.
# TYPE go_memstats_stack_sys_bytes gauge
go_memstats_stack_sys_bytes 1.376256e+06
# HELP go_memstats_sys_bytes Number of bytes obtained from system.
# TYPE go_memstats_sys_bytes gauge
go_memstats_sys_bytes 3.1544584e+07
# HELP go_threads Number of OS threads created.
# TYPE go_threads gauge
go_threads 12
# HELP kube_api_read_cache_hits The total number of read requests served from cache when looking up object metadata
# TYPE kube_api_read_cache_hits counter
kube_api_read_cache_hits 5
# HELP kube_api_read_cache_misses The total number of read requests served from kube-apiserver when looking up object metadata
# TYPE kube_api_read_cache_misses counter
kube_api_read_cache_misses 12
# HELP send_event_errors The total number of send event errors
# TYPE send_event_errors counter
send_event_errors 0
# HELP watch_errors The total number of errors received from the informer
# TYPE watch_errors counter
watch_errors 0
```

Reply:
The expectation is that it should only emit events that are currently happening on the resource. Did any other event happen after the Created or Scheduled events occurred on the pods?

Reply:
Yes, after the Scheduled Event i had also Events like Pulled, Created, Started. 
I thought that all Events are emitted that i've configured regardless of any Event that is emitted afterwards.
Thanks for the clarification.